File: src/statusbar/modules/ram.ts
import type {
  StatusBarModule,
  StatusBarModuleContext,
  Segment,
} from "../types.ts";

function formatMB(mb: number): string {
  if (mb >= 1024) return `${(mb / 1024).toFixed(1)}G`;
  return `${Math.round(mb)}M`;
}

export const ramModule: StatusBarModule = {
  id: "ram",
  render(ctx: StatusBarModuleContext): Segment[] {
    const mem = ctx.metrics.memory;
    const icon = ctx.icons ? "󰍛 " : "";
    const showDetail = ctx.moduleConfig.showDetail as boolean;

    let text: string;
    if (showDetail) {
      text = ` ${icon}${formatMB(mem.usedMB)}/${formatMB(mem.totalMB)} `;
    } else {
      text = ` ${icon}${Math.round(mem.percentage)}% `;
    }

    return [
      {
        text,
        fg: ctx.colors.fg,
        bg: ctx.colors.bg,
      },
    ];
  },
};

File: rust/crates/maxmux-statusbar/src/modules/ram.rs
use crate::modules::{config_bool, ModuleContext, StatusBarModule};
use crate::types::Segment;

pub struct RamModule;

/// Format megabytes: if >= 1024 show as GB with 1 decimal, else show as MB.
fn format_mb(mb: f64) -> String {
    if mb >= 1024.0 {
        format!("{:.1}G", mb / 1024.0)
    } else {
        format!("{:.0}M", mb)
    }
}

impl StatusBarModule for RamModule {
    fn id(&self) -> &str {
        "ram"
    }

    fn render(&self, ctx: &ModuleContext) -> Vec<Segment> {
        let Some(mem) = &ctx.metrics.memory else {
            return vec![];
        };

        let show_detail = config_bool(ctx, "showDetail", false);

        let content = if show_detail {
            format!("{}/{}", format_mb(mem.used_mb), format_mb(mem.total_mb))
        } else {
            format!("{:.0}%", mem.percentage)
        };

        let text = if ctx.icons {
            format!(" \u{f035b} {} ", content)
        } else {
            format!(" {} ", content)
        };

        vec![Segment::new(&text, &ctx.colors.fg, &ctx.colors.bg)]
    }
}
